Handle update of an exported service

blue-troy opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

When an exported service update, Service deletion
will trigger ServiceExport update.

Fixes #645
Signed-off-by: blue-troy <12729455+blue-troy@users.noreply.github.com>

tpantelis commented on pkg/agent/controller/agent.go:
```diff
@@ -381,6 +385,20 @@ func (a *Controller) serviceToRemoteServiceImport(obj runtime.Object, numRequeue
+		a.updateExportedServiceStatus(svcExport.Name, svcExport.Namespace, corev1.ConditionFalse, serviceUpdate,
+			"Service update")
+
+		return nil, false
```
What specific modification of an exported service would warrant this action? It seems potentially disruptive to delete the `ServiceImport` - why not update the existing one?

blue-troy replied:
The service may change to an unsupported type service. service Import seem don't have a not avalible status

tpantelis replied:
Then let's handle that specific case rather than blindly deleting the `ServiceImport`.

blue-troy replied:
Good idea, i will try to solve it in the next few days

blue-troy replied:
update `ServiceExport` status in Service deletion, let `ServiceExport -> ServiceImport` decide how to deal with it.

tpantelis commented on pkg/agent/controller/agent.go:
```diff
@@ -381,6 +391,13 @@ func (a *Controller) serviceToRemoteServiceImport(obj runtime.Object, numRequeue
 	// Update the status and requeue
+	if op == syncer.Update {
+		a.updateExportedServiceStatus(svcExport.Name, svcExport.Namespace, corev1.ConditionFalse, serviceUpdate,
+			"Service update")
```
Rather than changing the status condition to an intermediate state indicating it's not valid and relying on a `ServiceExport` update to trigger `serviceExportToServiceImport`, I think it would be cleaner to call `serviceExportToServiceImport`  directly:

`return serviceExportToServiceImport(svcExport, numRequeues, syncer.Update)`

blue-troy replied:
When call `return serviceExportToServiceImport(svcExport, numRequeues, syncer.Update)` directly

[serviceExportToServiceImport code](https://github.com/submariner-io/lighthouse/blob/v0.14.0-m1/pkg/agent/controller/agent.go#L263)

```go
	if op == syncer.Update && getLastExportConditionReason(svcExport) != serviceUnavailable {
		return nil, false
	}
```

LastExportConditionReason "" may be trigger by `onSuccessfulServiceImportSync` or `serviceToRemoteServiceImport` directly call at least. will ConditionReason "" may be somthing else? If it's not,we can change onSuccessfulServiceImportSync reason with `ServiceImportSynced`.

If we do not want to change `onSuccessfulServiceImportSync` reason, change `ServiceExport` then directly call `serviceToRemoteServiceImport` will cause `serviceToRemoteServiceImport` triggered more than it's needs.

tpantelis replied:
Sorry I'm not sure I follow. So if `serviceToRemoteServiceImport` is called directly then the above check will skip updating the `ServiceImport`? If so then perhaps instead of checking the last condition reason we check if the last condition Status is `ConditionTrue`, ie process the update only if the service was previously exported.

blue-troy replied:
you mean [serviceExportToServiceImport](https://github.com/submariner-io/lighthouse/blob/v0.14.0-m1/pkg/agent/controller/agent.go#L263) like this ?：

```go
	if op == syncer.Update && getLastExportCondition(svcExport) != corev1.ConditionTrue {
		return nil, false
	}
```

onSuccessfulServiceImportSync will trigger this ConditionTrue too. how distinguish service update direct call and onSuccessfulServiceImportSync? service update direct call and onSuccessfulServiceImportSync has the same Status and Reason

tpantelis replied:
https://github.com/submariner-io/lighthouse/pull/899 should alleviate your issues with calling `serviceToRemoteServiceImport` directly.

blue-troy replied:
OK, i will have a look

blue-troy replied:
#899 is useful in our case 👍, i will push direct call `serviceToRemoteServiceImport` version code after your  #899 PR merged

blue-troy committed:
Handle update an exported service

When an exported service update, Service deletion
will trigger ServiceExport update.

This commit handle case like service port numbers
update,but not handle service type update from a
supported type to an unsupported type.

Fixes #645
Signed-off-by: blue-troy <12729455+blue-troy@users.noreply.github.com>

blue-troy commented:
should we update [code](https://github.com/submariner-io/lighthouse/blob/8a8c1fbfc8d34b98c15aaef670d1c3797cdb2fbf/pkg/agent/controller/agent.go#L269-L277) like this ?:

```go
	svcType, ok := getServiceImportType(svc)

	if !ok {
		a.updateExportedServiceStatus(svcExport.Name, svcExport.Namespace, mcsv1a1.ServiceExportValid, corev1.ConditionFalse,
			invalidServiceType, fmt.Sprintf("Service of type %v not supported", svc.Spec.Type))
		klog.Errorf("Service type %q not supported", svc.Spec.Type)

		//delete unsupported service type
		err := a.serviceImportSyncer.GetLocalFederator().Delete(serviceImport)
                 ....
}
```

to support case like an exported service type change from a supported type to an unsupported type?

vthapar commented:
> should we update [code](https://github.com/submariner-io/lighthouse/blob/8a8c1fbfc8d34b98c15aaef670d1c3797cdb2fbf/pkg/agent/controller/agent.go#L269-L277) like this ?:
> 
> ```go
> 	svcType, ok := getServiceImportType(svc)
> 
> 	if !ok {
> 		a.updateExportedServiceStatus(svcExport.Name, svcExport.Namespace, mcsv1a1.ServiceExportValid, corev1.ConditionFalse,
> 			invalidServiceType, fmt.Sprintf("Service of type %v not supported", svc.Spec.Type))
> 		klog.Errorf("Service type %q not supported", svc.Spec.Type)
> 
> 		//delete unsupported service type
> 		err := a.serviceImportSyncer.GetLocalFederator().Delete(serviceImport)
>                  ....
> }
> ```
> 
> to support case like an exported service type change from a supported type to an unsupported type?

It should be done becasue a serviceImport shouldn't exist for a service that isn't supported as per current logic. But would like to add a check for `op.Update`

@tpantelis In case of delete of something that doesn't exist yet, do we silently ignore, or log it?

vthapar commented on pkg/agent/controller/agent.go:
```diff
@@ -273,7 +273,7 @@ func (a *Controller) serviceExportToServiceImport(obj runtime.Object, numRequeue
 		klog.Errorf("Service type %q not supported", svc.Spec.Type)
 
-		return nil, true
+		return nil, false
```
Why changing this to false?

blue-troy replied:
serviceToRemoteServiceImport can cover `service change from not supported type to supported type` case, by call a.serviceExportToServiceImport.

tpantelis replied:
It should. We should add unit test cases to cover the new changes. I can do that in a follow-up PR.

tpantelis commented:
> It should be done becasue a serviceImport shouldn't exist for a service that isn't supported as per current logic. But would like to add a check for `op.Update`
> 
> @tpantelis In case of delete of something that doesn't exist yet, do we silently ignore, or log it?

It doesn't: https://github.com/submariner-io/admiral/blob/e07dc24fda439c217cf8c512161474b1038aa355/pkg/federate/base_federator.go#L70. It's up to the caller to check the error.

tpantelis committed:
Delete ServiceImport when service type changes to unsupported

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Add unit test cases for Service updates

Also refactored awaitServiceExportCondition to use an informer/chan
to properly handle condition updates.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis commented:
OK I pushed 2 additional commits: the first deletes the `ServiceImport` when the type is unsupported and the second adds unit tests cases to cover the changes. I think we should get this in for 0.14. @vthapar @aswinsuryan 

**Edit**: also pushed commit for the release note.

tpantelis committed:
Add release note for service updates

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

dfarrell07 reviewed (approved):
LGTM, thanks @blue-troy!

skitt committed:
Merge branch 'devel' into fix/service_update

skitt commented:
@vthapar this is waiting on your re-review.

tpantelis committed:
Merge branch 'devel' into fix/service_update

vthapar reviewed (approved):
Thanks for all the work on this @blue-troy and @tpantelis . Looks good.
